File: go/grains/grains.go
package grains

import "errors"

const squares = 64

// Square returns how many grains are on the nTh square,
// where the first square has 1 and every square thereafter has double that.
// Could implement via math.Pow(2, n), but bitshifting is equivalent and much faster
func Square(n int) (uint64, error) {

	if n <= 0 || n > squares {
		return 0, errors.New("Square out of bounds")
	}

	return uint64(1 << uint(n-1)), nil
}

// Total returns the total amount of grains on a checkboard,
// where the first square has 1 and every square thereafter has double that.
func Total() uint64 {
	return uint64((1 << squares) - 1)
}

File: erlang/grains/src/grains.erl
-module(grains).

-export([square/1, total/0]).

%% How many grains are on Square N on a chessboard, where each square
%% contains double the previous square's amount of grains.
%% We could compute this recursively, but this happens to be power of
%% X series. We _could_ use bitshifting here, but let's keep it simple
%% for future devs to grok and use math:pow instead

square(N) when N < 1; N > 64 ->
    {error, "square must be between 1 and 64"};
square(N) -> trunc(math:pow(2, N - 1)).

%% As each tiles doubles the previous, the sum of all previous values
%% would be 1 less than the next (65th) theoretical tile
total() -> trunc(math:pow(2, 64)) - 1.
